src/core/lv2/Lv2Worker.cpp

```diff
@@ -0,0 +1,210 @@
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include <cassert>
+#include <QDebug>
+
+#include "Lv2Worker.h"
+
+#ifdef LMMS_HAVE_LV2
+
+#include "Engine.h"
+
+
+namespace lmms
+{
+
+
+// static wrappers
+
+static LV2_Worker_Status
+staticWorkerRespond(LV2_Worker_Respond_Handle handle,
+	uint32_t size, const void* data)
+{
+	Lv2Worker* worker = static_cast<Lv2Worker*>(handle);
+	return worker->respond(size, data);
+}
+
+
+
+
+std::size_t Lv2Worker::bufferSize() const
+{
+	// ardour uses this fixed size for ALSA:
+	return 8192 * 4;
+	// for jack, they use 4 * jack_port_type_get_buffer_size (..., JACK_DEFAULT_MIDI_TYPE)
+	// (possible extension for AudioDevice)
+}
+
+
+
+
+Lv2Worker::Lv2Worker(LV2_Handle handle,
+	const LV2_Worker_Interface* iface,
+	Semaphore* common_work_lock,
+	bool threaded) :
+	m_iface(iface),
+	m_threaded(threaded),
+	m_handle(handle),
+	m_response(bufferSize()),
+	m_requests(bufferSize()),
+	m_responses(bufferSize()),
+	m_requestsReader(m_requests),
+	m_responsesReader(m_responses),
+	m_sem(0),
+	m_workLock(common_work_lock)
+{
+	assert(iface);
+	m_scheduleFeature.handle = static_cast<LV2_Worker_Schedule_Handle>(this);
+	m_scheduleFeature.schedule_work = [](LV2_Worker_Schedule_Handle handle,
+		uint32_t size, const void* data) -> LV2_Worker_Status
+		{
+			Lv2Worker* worker = static_cast<Lv2Worker*>(handle);
+			return worker->scheduleWork(size, data);
+		};
+
+	if (threaded) { m_thread = std::thread(&Lv2Worker::workerFunc, this); }
+
+	m_requests.mlock();
+	m_responses.mlock();
+}
+
+
+
+
+Lv2Worker::~Lv2Worker()
+{
+	m_exit = true;
+	if(m_threaded) {
+		m_sem.post();
+		m_thread.join();
+	}
+}
+
+
+
+
+LV2_Worker_Status Lv2Worker::respond(uint32_t size, const void* data)
+{
+	if(!size) { return LV2_WORKER_ERR_UNKNOWN; }
+
+	LV2_Worker_Status rval = LV2_WORKER_SUCCESS;
+	if(m_threaded)
+	{
+		if(m_responses.free() < sizeof(size) + size)
+		{
+			rval = LV2_WORKER_ERR_NO_SPACE;
+		}
+		else
+		{
+			m_responses.write((const char*)&size, sizeof(size));
+			m_responses.write((const char*)data, size);
+		}
+	}
+	else
+	{
+		m_iface->work_response(m_handle, size, data);
+	}
+	return rval;
+}
+
+
+
+
+void Lv2Worker::workerFunc()
+{
+	std::vector<char> buf;
+	uint32_t size;
+	while (true) {
+		m_sem.wait();
+		if (m_exit) break;
+		const std::size_t readSpace = m_requestsReader.read_space();
+		if (readSpace <= sizeof(size)) continue; // (should not happen)
+
+		m_requestsReader.read(sizeof(size)).copy((char*)&size, sizeof(size));
+		assert(size <= readSpace - sizeof(size));
+		if(size > buf.size())
+			buf.resize(size);
+		m_requestsReader.read(size).copy(buf.data(), size);
+
+		m_workLock->wait();
+		m_iface->work(m_handle, staticWorkerRespond, this, size, buf.data());
+		m_workLock->post();
+	}
+}
+
+
+
+
+LV2_Worker_Status Lv2Worker::scheduleWork(uint32_t size, const void *data)
+{
+	if (!size) { return LV2_WORKER_ERR_UNKNOWN; }
```

Comment:
I can't see anything in the LV2 Worker documentation that suggests a size of zero is invalid. In fact, [`schedule_work`](https://lv2plug.in/c/html/group__worker.html#a48648b995e558a6cafeae29684c2ad93) explicitly indicates that `data` is allowed to be null, which is presumably paired with a size of zero (though it may be wise not to assume that, for stability).

Reply:
Wow, this is indeed dumb. I do not know why I added this restriction for 0-sizes. And in `Lv2Worker::respond` I did the same mistake...

While changing that, I also let the second write (of data) only be done if size and data are non-null, to cover a lot of potential issues and to save runtime with atomic variables.